autodocodec-api-usage/src/Autodocodec/Usage.hs

```diff
@@ -493,3 +494,81 @@ instance HasCodec MultilineDefault where
     object "MultilineDefault" $
       MultilineDefault
         <$> optionalFieldWithDefault "value" (Via "foo" "bar") "a field with a multi-line default value" .= multilineDefaultValue
+
+data These
+  = This Text
+  | That Int
+  | Both Text Int
+  deriving stock (Show, Eq, Generic)
+  deriving
+    ( FromJSON,
+      ToJSON,
+      Swagger.ToSchema,
+      OpenAPI.ToSchema
+    )
+    via (Autodocodec These)
+
+instance Validity These
+
+instance NFData These
+
+instance GenValid These where
+  genValid = genValidStructurallyWithoutExtraChecking
+  shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
+
+instance HasCodec These where
+  codec =
+    object "These" $
+      discriminatedUnionCodec "type" enc dec
+    where
+      textFieldCodec = requiredField' "text"
+      intFieldCodec = requiredField' "int"
+      bothFieldsCodec = (,) <$> textFieldCodec .= fst <*> intFieldCodec .= snd
+      enc = \case
+        This s -> ("this", SomeEncodable s textFieldCodec)
+        That n -> ("that", SomeEncodable n intFieldCodec)
+        Both s n -> ("both", SomeEncodable (s, n) bothFieldsCodec)
+      dec =
+        InsOrdHashMap.fromList
+          [ ("this", SomeDecodable textFieldCodec "This" This),
+            ("that", SomeDecodable intFieldCodec "That" That),
+            ("both", SomeDecodable bothFieldsCodec "Both" (uncurry Both))
+          ]
+
+data Expression
+  = LiteralExpression Int
+  | SumExpression Expression Expression
+  | ProductExpression Expression Expression
+  deriving stock (Show, Eq, Generic)
+  deriving
+    ( FromJSON,
+      ToJSON,
+      Swagger.ToSchema,
+      OpenAPI.ToSchema
+    )
+    via (Autodocodec Expression)
+
+instance Validity Expression
+
+instance NFData Expression
+
+instance GenValid Expression where
+  genValid = genValidStructurallyWithoutExtraChecking
```

Comment:
Generating generators for recursive types will lead to infinite generations. You have to write this manually.